libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/TelemetryWrapper.kt

```diff
@@ -0,0 +1,195 @@
+        observers.forEach {
+            it.onStateChanged(state.isEnabled)
+        }
+    }
+
+    fun registerObserver(observer: Observer) {
+        observers.add(observer)
+        if (observers.size == 1) {
+            telemetryUtils.registerTelemetryCollectionStateObserver(telemetryStateObserver)
+        }
+    }
+
+    fun unregisterObserver(observer: Observer) {
+        observers.remove(observer)
+        if (observers.isEmpty()) {
+            telemetryUtils.unregisterTelemetryCollectionStateObserver(telemetryStateObserver)
+        }
+    }
+
+    fun interface Observer {
+        fun onStateChanged(telemetryEnable: Boolean)
+    }
+
+    private companion object {
+        val TelemetryCollectionState.isEnabled: Boolean
+            get() = this == TelemetryCollectionState.ENABLED
+    }
+}
+
+/**
+ * Class that manages [MapboxNavigationTelemetry] initialisation. Listens to telemetry state events
+ * and can enable/disable telemetry in runtime.
+ *
+ * TODO(NAVAND-1820) ensure that we have 1-1 relationship between [TelemetryWrapper] and [MapboxNavigationTelemetry]
+ * [MapboxNavigationTelemetry] is very complex already and needs to be refactored.
+ */
+@UiThread
+internal class TelemetryWrapper(
+    private val telemetryStateWatcher: TelemetryStateWatcher = TelemetryStateWatcher()
+) {
+
+    private lateinit var mapboxNavigation: MapboxNavigation
+    private lateinit var navigationOptions: NavigationOptions
+    private lateinit var userAgent: String
+
+    private var isWrapperInitialized = false
+    private var isTelemetryEnabled = false
+
+    private val telemetryStateObserver = TelemetryStateWatcher.Observer { isEnabled ->
+        if (!isTelemetryEnabled && isEnabled) {
+            initializeSdkTelemetry()
+        } else if (isTelemetryEnabled && !isEnabled) {
+            unInitializeSdkTelemetry()
+        }
+    }
+
+    fun initialize(
+        mapboxNavigation: MapboxNavigation,
+        navigationOptions: NavigationOptions,
+        userAgent: String
+    ) {
+        check(!isWrapperInitialized) {
+            "Already initialized"
+        }
+        isWrapperInitialized = true
+
+        this.mapboxNavigation = mapboxNavigation
+        this.navigationOptions = navigationOptions
+        this.userAgent = userAgent
+
+        if (TelemetryUtilsDelegate.getEventsCollectionState()) {
+            initializeSdkTelemetry()
+        }
+
+        telemetryStateWatcher.registerObserver(telemetryStateObserver)
+    }
+
+    fun destroy() {
+        check(isWrapperInitialized) {
+            "Initialize object first"
+        }
+        isWrapperInitialized = false
+
+        telemetryStateWatcher.unregisterObserver(telemetryStateObserver)
+        if (isTelemetryEnabled) {
+            unInitializeSdkTelemetry()
+        }
+    }
+
+    fun postCustomEvent(
+        payload: String,
+        @CustomEvent.Type customEventType: String,
+        customEventVersion: String,
+    ) {
+        if (!isTelemetryEnabled) {
```

Comment:
I find it a bit inobvious that whenever you add a method here, you have to check the flag. I mean it's easy to forget to do that.
 I'd introduce an interface TelemetryEventsSender (not sure if sender is the best name here) and 2 implementations: EnabledTelemetryEventsSender and DisabledTelemetryEventsSender (or dummy, whatever).
This class (wrapper) has a var of type TelemetryEventsSender. And whenever the isTelemetryEnabled flag changes, we change the implementation correspondingly. WDYT?

Reply:
Yeah, I like this approach, wdyt if we improve that in the follow-up ticket NAVAND-1820?

Reply:
This ticket is P2 and called "Telemetry refactoring", we'll never get to that, especially considering that it's v2.
But now I find this approach with flags a bit difficult to maintain and error-prone and I'd vote changing it here, it also doesn't look like too much work.

Reply:
Ok, I've changed implementation